packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts

```diff
@@ -0,0 +1,514 @@
+          else return 'userProfile email is undefined';
+        } //if
+        else return 'userProfile is undefined';
+      }
+    }
+
+    app.controller(InfoController);
+
+    const invalidToken = 'aaa.bbb.ccc.ddd';
+
+    await whenIMakeRequestTo(server)
+      .get('/whoAmI')
+      .set('Authorization', 'Bearer ' + invalidToken)
+      .expect(401);
+  });
+
+  it('create a json web token throws error for missing email', async () => {
+    class InfoController {
+      constructor(
+        @inject(JWTAuthenticationStrategyBindings.TOKEN_SERVICE)
+        public tokenService: JWTService,
+        @inject(USER_REPO)
+        public users: UserRepository,
+      ) {}
+
+      @get('/createtoken')
+      async createToken() {
+        const joeUser = this.users.list['joe@example.com'].user;
+
+        const joeUserProfile = {
+          id: joeUser.id,
+          email: undefined,
+          name: `${joeUser.firstname} ${joeUser.surname}`,
+        };
+
+        return await this.tokenService.generateToken(joeUserProfile);
+      }
+    }
+
+    app.controller(InfoController);
+
+    await whenIMakeRequestTo(server)
+      .get('/createtoken')
+      .expect(401);
+  });
+
+  it('create a json web token throws error for missing name', async () => {
+    class InfoController {
+      constructor(
+        @inject(JWTAuthenticationStrategyBindings.TOKEN_SERVICE)
+        public tokenService: JWTService,
+        @inject(USER_REPO)
+        public users: UserRepository,
+      ) {}
+
+      @get('/createtoken')
+      async createToken() {
+        const joeUser = this.users.list['joe@example.com'].user;
+
+        const joeUserProfile = {
+          id: joeUser.id,
+          email: joeUser.email,
+          name: undefined,
+        };
+
+        return await this.tokenService.generateToken(joeUserProfile);
+      }
+    }
+
+    app.controller(InfoController);
+
+    await whenIMakeRequestTo(server)
+      .get('/createtoken')
+      .expect(401);
+  });
+
+  it('allows anonymous requests to methods with no decorator', async () => {
+    class InfoController {
+      @get('/status')
+      status() {
+        return {running: true};
+      }
+    }
+
+    app.controller(InfoController);
+    await whenIMakeRequestTo(server)
+      .get('/status')
+      .expect(200, {running: true});
+  });
+
+  it('returns error for unknown authentication strategy', async () => {
+    class InfoController {
+      @get('/status')
+      @authenticate('doesnotexist')
+      status() {
+        return {running: true};
+      }
+    }
+
+    app.controller(InfoController);
+    await whenIMakeRequestTo(server)
+      .get('/status')
+      .expect(401);
+  });
+
+  async function givenAServer() {
+    app = new Application();
+    app.component(AuthenticationComponent);
+    app.component(RestComponent);
+    server = await app.getServer(RestServer);
+  }
+
+  function givenAuthenticatedSequence() {
```

Comment:
Can we extract these functions into a separate `helper` ts file so that it can be shared by multiple tests?

Reply:
@raymondfeng , please specify which functions should be extracted and re-used by which tests.   I guess `getAuthenticatedSequence()`  and `givenAServer()` and the `user repo` are common for `packages/authentication/src/__tests__/acceptance/basic-auth-extension.acceptance.ts` and `packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts`.